### DISCUSSION TITLE
Set (Eco) Temperature before turning off climate underlying

### DISCUSSION THREAD

**Discussion by andrexp** [jmcollin78/versatile_thermostat#642] [Q&A] [answered]
Hi,

I look for a solution for a special case. I use the LCN integration for HA to control my thermostats. The integration publishes climate-entities. To utilize the versatile thermostat features I use a template to extract the current temperature from the rooms. So far so good. I also use the window-open-feature (by binary sensors) to power down the climate(s) when windows were opened. Unfortunately the LCN integration "locks" the regulator value to the last value when "turn off" is sent. This results unfortunately in locking the setpoint temperature to 20 °C for example. So if the temperature drops because of the opened window the thermostat will try to hold the temperature and starts to heat.

My first thought was simply use the versatile thermostat function setting the eco temperature in case of an opened window. But then setpoint changes lead to overwrite the reduced temperature. So I would need to add an additional temperature set (preferably the eco temperature) before turning off (lock my thermostats).

Any suggestion where I could add this?

**Comment by andrexp** [comment, marked as answer]
After more research I found a solution for me. Adding the following to line 2524 of base_thermostat.py seems to do the trick:

```
            else:  # default is to turn_off
                # send eco-temperature before turn off climate
                for under in self._underlyings:
                    await under.set_temperature(
                        self.find_preset_temp(PRESET_ECO),
                        self._attr_max_temp,
                        self._attr_min_temp,
                    )
                self.set_hvac_off_reason(HVAC_OFF_REASON_WINDOW_DETECTION)
                await self.async_set_hvac_mode(HVACMode.OFF)
```

I simply want to send the (eco-)temperature (this is only the proof of concept) to the underlying climate without changing the versatile thermostats temperature itself. This leads into changing the temperature to a lower one and locking the versatile thermostat in off-state. Now it is possible to change the temperature (for example through an automation rule) without being in danger that the heating is turned on while the window is still open. When the window is closed again, either the previously set temperature or the newly set temperature through any automation rule is set.

I am aware that this is a quick and dirty hack. Maybe there is a more elegant way to implement this (e.g. by adding a selection if and which temperature should be sent)? I hope this could also be a use case to other users.

**Comment by jmcollin78** [comment]
Hello @andrexp ,

I'm confused. This feature already exists natively in Vtherm. When window open is detected you can turn off or change the preset.

You just have to configure what is the action when window is detected as open.

**Comment by andrexp** [comment]
Hi,

thank you for your answer. I knew that. As I wrote above:

> My first thought was simply use the versatile thermostat function setting the eco temperature in case of an opened window. But then setpoint changes lead to overwrite the reduced temperature. So I would need to add an additional temperature set (preferably the eco temperature) before turning off (lock my thermostats).
> 

Unfortunately this solution is not what I need. I also control my climates manually and through rules. Reducing the temperature with the native VTherm feature does not block temperature changes and it could be that someone changes the temperature when window is still opened. Please correct me if I am using it wrong.

**Reply by jmcollin78** [reply]
You want to send a preset temp AND to turn-off the VTherm ?
Is that your point ?

> This results unfortunately in locking the setpoint temperature to 20 °C for example.

Did you mean that when off is send to the underlying the off command is ignored and the device continue to heat with the last value ?

If yes this means you cannot use the turn-off when window is detected, and use the "send preset" when window is detected.

**Comment by andrexp** [comment]
Yes, that's my problem. The underlying interprets the "turn-off" in locking the currently set Temperature. That means if you want to change the temperature by hand (I have displays in my room) it won't accept any setpoint changes. But if the thermostat locks to the last value it also can be 20 °C (or something higher) which is far too high for an opened window... :-)

**Reply by jmcollin78** [reply]
First time I see that. What is brand and model of your device ?

Have you tried to configure the VTherm to change the preset instead of turning off upon window detection ? It should fix your case.

**Comment by andrexp** [comment]
It is LCN (a German manufactured building automation system by the Issendorff KG).

Yes, I did that. But then setpoint values will have instant impact to underlyings if someone try to change it while window is open. While the VTherm is in turned-off state the setpoint change will be transmitted to the underlying when window is closed again.

**Reply by jmcollin78** [reply]
> But then setpoint values will have instant impact to underlyings if someone try to change it while window is open

Yes but I guess this is not a very common use case.

Your case is too much specific, I don't know what I can do and I will not change the behavior of VTherm for this very specific case.
Sorry for you.

**Comment by andrexp** [comment]
I basically agree with you, but I am wondering if sending a specified preset temperature to the underlying before HVACMode.OFF is set would change the general behaviour of VTerm. The thermostats would do that in a similar way anyway by closing the valve of the radiator or circuit completely, aren't they?

However, thank you for your time and for this great work.

**Reply by jmcollin78** [reply]
> if sending a specified preset temperature to the underlying before HVACMode.OFF is set would change the general behaviour of VTerm

by definition yes. And it is not so easy. Which preset ? What if this preset is deactivated ? What about the current preset (it will be lose when the thermostat turn on or I have to send a preset which is not stored), do we need a delay between the Eco preset and the shutdown.  I'm pretty sure this will generate unwanted side effect on AC system or heat pump.

> The thermostats would do that in a similar way anyway by closing the valve of the radiator or circuit completely, aren't they?

Each equipement does its own procedure to shutdown yes. Store the current setpoint, close the flap, ... I'm not sure it is a good idea to interfere with this procedure.

You better have to ask LCN why this weird behavior or change your system foro something more compliant.
